### DISCUSSION TITLE
Implementing support for a new RTOS

### DISCUSSION THREAD

**Discussion by Spivoxity** [pyocd/pyOCD#1135] [Q&A]
Is there a written guide anywhere to the steps needed to add support for a new RTOS to pyocd?  Code for several existing RTOS's is available, but I haven't found documentation for the interface it is implementing.

A simplified OS (tentatively called micro:bian) for the BBC micro:bit, based on message passing and reminiscent of Andy Tanenbaum's MINIX kernel, will be covered in my book "Bare Metal micro:bit".  I'd like to include debugger support if possible.

Best wishes,
-- Mike Spivey

**Comment by newbrain** [comment]
Having recently added the support for Azure RTOS/ThreadX (#1064), I can confidently say that there is no written guide.

The whole thing can be summed up in the following steps (if I find the time, I might try to make a PR with an expanded version of this:

* The structure is based on the specialization of a number of base classes
    * `Plugin`
    Simply defines the name, description and entry point (ThreadProvider) of the plug-in
    * `ThreadProvider`
    This class is responsible to make sure we are looking at an actual RTOS instance (in the `init` method) by cecking for RTOS specific symbols in the elf file.
    It also must be able to enumerate the threads, by walking the RTOS data structures (`_build_thread_list`, `get_thread`, `get_threads`)
    * `TargetThread`
    Represent a thread in the RTOS, knows about its state and how to get to its registers, using the specialized `DebugContext`
    * `DebugContext`
    The derived class understands the structure of a thread stack frame, and can return any register in it (on in the base context) with the `read_register_raw` method.
 * The plug-in needs to be registered in the top level `setup.py` file in order to be loaded (search for the `entry_points` dictionary)

I started mostly from the freeRTOS and argon plug-ins, taking liberal inspiration (AKA CtrlC Ctrl-V), then simplifying as much as possible.

**Comment by flit** [comment]
micro:bian sounds very cool! I'm a fan of small OSes and OS architecture, and I help out the micro:bit Foundation (for DAPLink) so it's fantastic to hear it being used for such work and educational resources.

As @newbrain stated, there's not a guide for writing an RTOS plugin, sorry about that. If you have any questions, feel free to ask here. I will try to extract a guide from what he's already written and anything I add here, with some details filled in.
